dali/pipeline/operators/reader/loader/lmdb.h

```diff
@@ -30,65 +31,147 @@ namespace dali {
+    } else if (index == mdb_index_ + 1) {
+      CHECK_LMDB(mdb_cursor_get(mdb_cursor_, key, value, MDB_NEXT), db_path_);
+    } else if (index > mdb_index_) {
+      LOG_LINE << "lmdb " << num_ << " " << db_path_
+               << " exec a large step forward " << mdb_index_ << "->" << index << std::endl;
+      for (Index i = mdb_index_; i < index; ++i) {
+        CHECK_LMDB(mdb_cursor_get(mdb_cursor_, key, value, MDB_NEXT), db_path_);
+      }
     } else {
-      CHECK_LMDB(status, filename);
-      return true;
+      // index < mdb_index_
+      LOG_LINE << "lmdb " << num_ << " " << db_path_
+               << " exec a large step backward " << mdb_index_ << "->" << index << std::endl;
+      for (Index i = index; i < mdb_index_; i++) {
+        CHECK_LMDB(mdb_cursor_get(mdb_cursor_, key, value, MDB_PREV), db_path_);
+      }
     }
+    mdb_index_ = index;
   }
 
-  inline uint64_t LMDB_size(MDB_txn* txn, MDB_dbi dbi, const string &filename) {
-    MDB_stat stat;
-
-    CHECK_LMDB(mdb_stat(txn, dbi, &stat), filename);
-
-    return stat.ms_entries;
+  void Close() {
+    if (mdb_cursor_) {
+      mdb_cursor_close(mdb_cursor_);
+      mdb_dbi_close(mdb_env_, mdb_dbi_);
+      mdb_cursor_ = nullptr;
+    }
+    if (mdb_transaction_) {
+      mdb_txn_abort(mdb_transaction_);
+      mdb_transaction_ = nullptr;
+    }
+    if (mdb_env_) {
+      mdb_env_close(mdb_env_);
+      mdb_env_ = nullptr;
+    }
   }
+};
 
-  inline void PrintLMDBStats(MDB_txn* txn, MDB_dbi dbi, const string &filename) {
-    MDB_stat stat;
-
-    CHECK_LMDB(mdb_stat(txn, dbi, &stat), filename);
+static int find_lower_bound(const std::vector<Index>& a, Index x) {
+  DALI_ENFORCE(x >= a.front() && x < a.back() && a.size() >= 2);
+  int low = 0;
+  int high = a.size()-2;
+  do {
+    int mid = (low+high) / 2;
+    if (x >= a[mid] && x < a[mid+1]) {
+      return mid;
+    } else if (x >= a[mid+1]) {
+      low = mid+1;
+    } else if (x < a[mid]) {
+      high = mid-1;
+    }
+  } while (low <= high);
 
-    printf("DB has %d entries\n", static_cast<int>(stat.ms_entries));
-  }
-}  // namespace lmdb
+  DALI_FAIL("the array is not in ascending order.");
+  return -1;
+}
 
 class LMDBLoader : public Loader<CPUBackend, Tensor<CPUBackend>> {
  public:
   explicit LMDBLoader(const OpSpec& options)
-    : Loader(options),
-      db_path_(options.GetArgument<string>("path")) {
-  }
+      : Loader(options),
+        db_paths_(options.GetRepeatedArgument<std::string>("path")) {}
 
   ~LMDBLoader() override {
-    if (mdb_cursor_) {
-      mdb_cursor_close(mdb_cursor_);
-      mdb_dbi_close(mdb_env_, mdb_dbi_);
+    for (size_t i = 0; i < db_paths_.size(); i++) {
+      mdb_[i].Close();
     }
-    if (mdb_transaction_)
-      mdb_txn_abort(mdb_transaction_);
-    if (mdb_env_)
-      mdb_env_close(mdb_env_);
-    mdb_env_ = nullptr;
+  }
+
+  void MapIndexToFile(Index index, Index* file_index, Index* local_index) {
```

Comment:
Although it is not used here it is possible to use this code with any form of random access to lmdb. The question is if it is needed?
Maybe we can assume that we just read sequentially entries from the first path, db run out of samples we move to the second path, and at the end, we wrap to the first path back.

Reply:
Thank you for your review. I have implemented a RandomLMDBLoader as a DALI plugin and used it for my caffe detection project. 
It reads from a random dbpath at first, when run out of samples it will randomly chooses the next dbpath to read. 
If it is needed I could add this to the PR.

Reply:
Random selection of path to reading next won't have any perf implication, I only worry about reading any sample randomly from any db.
If you want to add such a feature you need to make sure that shards doesn't overlap and are consistent between DALI instances running on the different GPUs.

Reply:
It it hard to conform to the shard mechanism for RandomLMDBLoader without any perf implication, it is use for single GPU only and is not suitable for this PR.